lib/dma.c

```diff
@@ -167,7 +167,13 @@ MOCK_DEFINE(dma_controller_remove_region)(dma_controller_t *dma,
         }
 
         if (dma_unregister != NULL) {
+#ifdef DEBUG
```

Comment:
seems useful, but:

 - no reason to be debug only
 - we could track for debugging purposes which cb we're in (make it an enum)
 - let's add an inline helper in private.h or something

Reply:
> no reason to be debug only
We document this, do we need to police the user of the library for breaking contract?

> we could track for debugging purposes which cb we're in (make it an enum)
If we keep this in debug, the asserts will produce a stack trace so it should be clear which cb we're in.

> let's add an inline helper in private.h or something
Not sure I understand, are you referring to this?
>  if you think it's useful then we could introduce a function through which we'execute callbacks (the callback will be passed as an argument to that function)

Reply:
> asserts will produce a stack trace

Only showing vfu_map_sg() etc. It won't show other situations where we are in callbacks. I'm saying, since you're tracking that user is in a callback, why not add a little extra  info so we know which one?

Reply:
>inline helper
I'm saying instead of open coding the assert, but maybe that's not really necessary

Reply:
> Only showing vfu_map_sg() etc. It won't show other situations where we are in callbacks.

good point, we do need the extra info you suggest

Reply:
Just realized I misunderstood something:
> we could track for debugging purposes which cb we're in (make it an enum)

These functions are allowed to be called even in quiesced state provided that they're called by a callback, so why track which callback we're in? If the call to e.g. `vfu_map_sg()` is not allowed, then it's guaranteed that it won't be call by a callback. Do you want the callback type printed each type we call these type of functions?

Reply:
From this:

```
  vfu_ctx->in_cb = true;
```

to this:

```
  vfu_ctx->in_cb = VFU_CTX_CB_DMA_UNREGISTER;
```

Reply:
I get that, my point is why track the callback type? How are we going to use it? Accessing function like `vfu_map_sg()` and friends is allowed in done by a callback.

Reply:
just for debugging purposes if we have a core. it's no big deal.